### ISSUE TITLE
pyarrow conversion logic should handle numeric columns with some string number values

### ISSUE THREAD

**Issue by jowens** [vega/vegafusion#472]
Versions: Python 3.12.2 (MacPorts), vegafusion / vegafusion-python-embed 1.6.5 (pip), apache-arrow 13.0.0_7 (MacPorts), py312-arrow 1.3.0_1 (MacPorts), py312-pyarrow 13.0.0_7 (MacPorts), py312-altair 5.2.0_0 (MacPorts), py312-pandas @2.2.1_0 (MacPorts)

I have a giant messy DataFrame with tons of columns, and I make several graphs from it. vf's ability to remove extraneous columns is fantastic.

I have been successfully using json.dumps to dump a JSON into an HTML file and make a .html version of an Altair graph. Some of these save perfectly well with vf.save_html. But I don't know how to debug it when they don't. The trace appears to indicate that my Pandas DataFrame doesn't serialize nicely into arrow, but I don't understand the error message (`pyarrow.lib.ArrowInvalid: Could not convert '12.0' with type str: tried to convert to double`, shouldn't `12.0` convert to double just fine? It would be nice to be able to identify the item that's failing (e.g., column name, row number). But first, I'd like to understand what this error message actually means.

```
    vf.save_html(chart, os.path.join(outputdir, title + "_vf.html"))
  File "/Users/jowens/Library/Python/3.12/lib/python/site-packages/vegafusion/save.py", line 40, in save_html
    bundle = spec_to_mime_bundle(
             ^^^^^^^^^^^^^^^^^^^^
  File "/Users/jowens/Library/Python/3.12/lib/python/site-packages/vegafusion/renderer.py", line 40, in spec_to_mime_bundle
    tx_vega_spec, warnings = runtime.pre_transform_spec(
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jowens/Library/Python/3.12/lib/python/site-packages/vegafusion/runtime.py", line 360, in pre_transform_spec
    imported_inline_dataset = self._import_or_register_inline_datasets(inline_datasets)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jowens/Library/Python/3.12/lib/python/site-packages/vegafusion/runtime.py", line 231, in _import_or_register_inline_datasets
    imported_inline_datasets[name] = PandasDatasource(value)
                                     ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jowens/Library/Python/3.12/lib/python/site-packages/vegafusion/datasource/pandas_datasource.py", line 31, in __init__
    candidate_schema = pa.Schema.from_pandas(df.iloc[::sample_stride][[col]])
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pyarrow/types.pxi", line 2621, in pyarrow.lib.Schema.from_pandas
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/pyarrow/pandas_compat.py", line 551, in dataframe_to_types
    type_ = pa.array(c, from_pandas=True).type
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pyarrow/array.pxi", line 323, in pyarrow.lib.array
  File "pyarrow/array.pxi", line 83, in pyarrow.lib._ndarray_to_array
  File "pyarrow/error.pxi", line 100, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: Could not convert '12.0' with type str: tried to convert to double
```

**Comment by jonmmease** [comment]
Hi @jowens, I'm glad VegaFusion has been helpful!

PyArrow is pretty strict about type coercion, and if a column has a mix of strings a numbers it won't try to coerce the strings to numbers (or the numbers to strings, which is also what you want sometimes).  Right now we do have some special logic to try to perform manual casts when PyArrow fails, but it looks like we're not catching the `ArrowInvalid` exception (only the `ArrowTypeError` exception).  But even if that logic was triggered, it would convert the entire column to strings, which is probably not what you want here.

When we get this error from pyarrow, I think it would make sense to attempt to convert the column to numbers before falling back to converting it to strings.  And then if any other exception is raised, we should wrap it to report the column the failed. Getting the row number might be doable, but would require parsing the `ArrowInvalid` error message to extract the value to search for and so could be pretty fragile to changes in the error message.

If you're interested in trying this, take a look at https://github.com/hex-inc/vegafusion/blob/3615368b49dced1f42770c32a70d06ba54f99047/python/vegafusion/vegafusion/transformer.py#L31-L45

In the near term, here's a workaround.

```python
# Imports
import pyarrow as pa
import pandas as pd
import vegafusion as vf

# Build DataFrame with mixed type column
df = pd.DataFrame({"a": [1.0, 2.0, "3.0"]})
vf.transformer.to_arrow_table(df)
```
```
...
ArrowInvalid: Could not convert '3.0' with type str: tried to convert to double
```
Find object columns with bad value:
```python
search_val = '3.0'
for col in df.columns:
    try:
        if df[col].dtype.kind == 'O':
            idx = df[col].tolist().index(search_val)
            print(f"col: {col}, index: {idx}")
    except ValueError:
        pass
```
```
col: a, index: 2
```
Cast bad column to float and perform conversion again
```python
df['a'] = df['a'].astype(float)
vf.transformer.to_arrow_table(df)
```
```
pyarrow.Table
a: double
b: string
----
a: [[1,2,3]]
b: [["a","b","3"]]
```

Hope that helps!

**Comment by jowens** [comment]
What a thoughtful answer. Thank you!

Just so I understand the error message, then: in my case "12.0" is a string in a mixed column and VF/Arrow have decided "That's not a double! I expected a double!" and the "convert to double" isn't actually trying to convert anything? 

```
df = pd.DataFrame({"a": [1.0, 2.0, "3.0"]})
vf.transformer.to_arrow_table(df)
...
ArrowInvalid: Could not convert '3.0' with type str: tried to convert to double
```

Simply understanding the error message might allow me to fix this issue before it gets to this point.

**Comment by jowens** [comment]
So with your suggestion, I am definitely able to find problematic cells that are preventing Arrowfication. This is helpful. So, as background:

I have hundreds of columns in my DataFrame (it's an ingest of dozens and dozens of Wikipedia tables, all concatted together and it's far from clean data or column names). Most of those columns are not used. What I have done to date is simply to output Altair plots with the columns that I actually do use. This produces large JSONs and thus large .html output plots, so you know the problem with that. But at least I don't have to do any manual selection of my columns of interest.

The errors that I'm having are in columns that I don't care about. So, either I could make a manual list of the dozens of columns that I do care about and filter out the others (and I haven't done that to date), or I can make a manual list of the hundreds of columns I don't care about and filter them out (and I haven't done that either). In particular these lists break every single time some random Wikipedia user decides to put in a new column into one table. What I kind of hoped `vf.save_html` would do was FIRST decide the relevant columns THEN filter then out THEN Arrowify them, but it appears the Arrowification happens first. Am I reading this wrong?

**Comment by jonmmease** [comment]
Hi @jowens,

> What I kind of hoped vf.save_html would do was FIRST decide the relevant columns THEN filter then out THEN Arrowify them, but it appears the Arrowification happens first. Am I reading this wrong?

It almost works this way. VegaFusion does only convert the columns that are used into full Arrow tables, but in does need to determine the types of all of the columns up front. Looking at your original stack trace more carefully I see that this is error is actually coming from the step where we try to infer the types of all of the columns.  We should be able to make this more robust and fall back to reporting a string for messy columns. I'll try to take a look at this soon!

**Comment by jowens** [comment]
Oh! That would be splendid!

Um, is there an easy two-liner for "filter out all messy columns"? Filter on `df[col].dtype.kind == 'O':` ?

**Comment by jonmmease** [comment]
It was actually really straightforward, PR in progress in https://github.com/vega/vegafusion/pull/473

> is there an easy two-liner for "filter out all messy columns"? Filter on df[col].dtype.kind == 'O'

Unfortunately this would also filter out string columns, which pandas represents as 'object' columns, unless you use the [pyarrow integration](https://pandas.pydata.org/docs/user_guide/pyarrow.html). But I expect you'd run into the same pyarrow error in this case.

**Comment by jonmmease** [comment]
Fix released in 1.6.6
